ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Tables.hs

```diff
@@ -0,0 +1,596 @@
+  => (forall k v.
+          (Ord k, Eq v)
+       => mk1 k v
+       -> mk2 k v
+       -> mk3 k v
+     )
+  ->              l mk1
+  -> LedgerTables l mk2
+  ->              l mk3
+zipOverLedgerTables f l tables2 =
+    overLedgerTables
+      (\tables1 -> zipLedgerTables f tables1 tables2)
+      l
+
+type HasTickedLedgerTables :: LedgerStateKind -> Constraint
+class HasLedgerTables l => HasTickedLedgerTables l where
+  -- | NOTE: The 'IsMapKind' constraint is here for the same reason
+  -- it's on 'projectLedgerTables'
+  projectLedgerTablesTicked :: IsMapKind mk => Ticked1 l mk  -> LedgerTables l mk
+  default projectLedgerTablesTicked ::
+       LedgerTablesAreTrivial l
+    => Ticked1 l mk
+    -> LedgerTables l mk
+  projectLedgerTablesTicked _ = trivialLedgerTables
+
+  -- | NOTE: The 'IsMapKind' constraint is here for the same reason
+  -- it's on 'withLedgerTables'
+  withLedgerTablesTicked :: IsMapKind mk => Ticked1 l any -> LedgerTables l mk -> Ticked1 l mk
+
+overLedgerTablesTicked ::
+     (HasTickedLedgerTables l, IsMapKind mk1, IsMapKind mk2)
+  => (LedgerTables l mk1 -> LedgerTables l mk2)
+  -> Ticked1 l mk1
+  -> Ticked1 l mk2
+overLedgerTablesTicked f l =
+    withLedgerTablesTicked l $ f $ projectLedgerTablesTicked l
+
+mapOverLedgerTablesTicked ::
+     (HasTickedLedgerTables l, IsMapKind mk1, IsMapKind mk2)
+  => (forall k v.
+         (Ord k, Eq v)
+      => mk1 k v
+      -> mk2 k v
+     )
+  -> Ticked1 l mk1
+  -> Ticked1 l mk2
+mapOverLedgerTablesTicked f = overLedgerTablesTicked $ mapLedgerTables f
+
+zipOverLedgerTablesTicked ::
+     (HasTickedLedgerTables l, IsMapKind mk1, IsMapKind mk3)
+  => (forall k v.
+         (Ord k, Eq v)
+      => mk1 k v
+      -> mk2 k v
+      -> mk3 k v
+     )
+  -> Ticked1      l mk1
+  -> LedgerTables l mk2
+  -> Ticked1      l mk3
+zipOverLedgerTablesTicked f l tables2 =
+    overLedgerTablesTicked
+      (\tables1 -> zipLedgerTables f tables1 tables2)
+      l
+
+-- | LedgerTables are projections of data from a LedgerState and as such they
+-- can be injected back into a LedgerState. This is necessary because the Ledger
+-- rules are ignorant to UTxO-HD changes so by stowing the ledger tables, we are
+-- able to provide a Ledger State with a restricted UTxO set that is enough to
+-- execute the Ledger rules.
+type CanStowLedgerTables :: LedgerStateKind -> Constraint
+class CanStowLedgerTables l where
+
+  stowLedgerTables     :: l ValuesMK -> l EmptyMK
+  default stowLedgerTables :: LedgerTablesAreTrivial l => l ValuesMK -> l EmptyMK
+  stowLedgerTables = convertMapKind
+
+  unstowLedgerTables   :: l EmptyMK  -> l ValuesMK
+  default unstowLedgerTables :: LedgerTablesAreTrivial l => l EmptyMK -> l ValuesMK
+  unstowLedgerTables = convertMapKind
+
+{-------------------------------------------------------------------------------
+  Concrete ledger tables
+
+  We provide here  the usual structures that we expect to deal with, namely:
+  - @'ValuesMK'@: a map of keys to values
```

Comment:
I think this `@'` is unnecessary and adds noise to the source-code. I think `haddock` turns the identifiers into inline code (instead of links) if they are not in the scope of the current module.